File: src/Manipulators/Thumbnail.php
<?php

namespace Weserv\Images\Manipulators;

use Jcupitt\Vips\Image;
use Jcupitt\Vips\Intent;
use Jcupitt\Vips\Interpretation;
use Jcupitt\Vips\Size;
use Weserv\Images\Exception\ImageTooLargeException;
use Weserv\Images\Manipulators\Helpers\Utils;

/**
 * @property string $t
 * @property string|int $w
 * @property string|int $h
 * @property string $dpr
 * @property string $a
 * @property string $tmpFileName
 * @property string $page
 * @property int $rotation
 * @property bool $flip
 * @property bool $flop
 * @property string|bool $trim
 * @property string $gam
 *
 * @SuppressWarnings(PHPMD.ExcessiveClassComplexity)
 */
class Thumbnail extends BaseManipulator
{
    /**
     * Maximum image size in pixels.
     */
    protected ?int $maxImageSize;

    /**
     * Profile map to ensure that we use a device-
     * independent color space for the images we process.
     */
    protected array $profileMap = [
        Interpretation::SRGB => 'srgb',
        Interpretation::CMYK => 'cmyk'
    ];

    protected const VIPS_MAX_COORD = 10000000;

    /**
     * Create Thumbnail instance.
     *
     * @param int|null $maxImageSize Maximum image size in pixels.
     */
    public function __construct(?int $maxImageSize = null)
    {
        $this->maxImageSize = $maxImageSize;
    }

    /**
     * Get the maximum image size.
     *
     * @return int|null Maximum image size in pixels.
     */
    public function getMaxImageSize(): ?int
    {
        return $this->maxImageSize;
    }

    /**
     * Set the maximum image size.
     *
     * @param int|null $maxImageSize Maximum image size in pixels.
     */
    public function setMaxImageSize(?int $maxImageSize): void
    {
        $this->maxImageSize = $maxImageSize;
    }

    /**
     * Perform thumbnail image manipulation.
     *
     * @param Image $image The source image.
     *
     * @throws \Jcupitt\Vips\Exception
     * @throws ImageTooLargeException if the provided image is too large for processing.
     *
     * @return Image The manipulated image.
     */
    public function run(Image $image): Image
    {
        // Cast any strings to integers
        $this->w = (int)$this->w;
        $this->h = (int)$this->h;

        // Width and height should never be less than zero
        if ($this->w < 0) {
            $this->w = 0;
        }
        if ($this->h < 0) {
            $this->h = 0;
        }

        [$this->w, $this->h] = $this->applyDpr($this->w, $this->h, $this->getDpr());

        // Check if image size is greater then the maximum allowed image size after dimension is resolved
        $this->checkImageSize($image, $this->w, $this->h);

        return $this->doThumbnail($image, $this->getFit(), $this->w, $this->h);
    }

    /**
     * Indicating if we should not enlarge the output if the input width
     * *and* height are already less than the required dimensions
     *
     * @param string $fit The resolved fit.
     *
     * @return bool
     */
    public function withoutEnlargement(string $fit): bool
    {
        return $fit === 'fit' || $fit === 'squaredown';
    }

    /**
     * Resolve fit.
     *
     * @return string The resolved fit.
     */
    public function getFit(): string
    {
        if ($this->t === 'fit' ||
            $this->t === 'fitup' ||
            $this->t === 'square' ||
            $this->t === 'squaredown' ||
            $this->t === 'absolute' ||
            $this->t === 'letterbox'
        ) {
            return $this->t;
        }

        if ($this->t !== null && strpos($this->t, 'crop') === 0) {
            return 'crop';
        }

        return 'fit';
    }

    /**
     * Resolve the device pixel ratio.
     *
     * @return float The device pixel ratio.
     */
    public function getDpr(): float
    {
        if (!is_numeric($this->dpr)) {
            return 1.0;
        }

        if ($this->dpr < 0 || $this->dpr > 8) {
            return 1.0;
        }

        return (float)$this->dpr;
    }

    /**
     * Apply the device pixel ratio.
     *
     * @param int $width The target image width.
     * @param int $height The target image height.
     * @param float $dpr The device pixel ratio.
     *
     * @return int[] The modified width and height.
     */
    public function applyDpr(int $width, int $height, float $dpr): array
    {
        $width *= $dpr;
        $height *= $dpr;

        return [
            (int)$width,
            (int)$height,
        ];
    }

    /**
     * Check if image size is greater then the maximum allowed image size.
     *
     * @param Image $image The source image.
     * @param int $width The image width.
     * @param int $height The image height.
     *
     * @throws ImageTooLargeException if the provided image is too large for processing.
     *
     * @return void
     */
    public function checkImageSize(Image $image, int $width, int $height): void
    {
        if ($width === 0 && $height === 0) {
            $width = $image->width;
            $height = $image->height;
        }
        if ($width !== 0) {
            $width = (int)($height * ($image->width / $image->height));
        }
        if ($height !== 0) {
            $height = (int)($width / ($image->width / $image->height));
        }

        if ($this->maxImageSize !== null) {
            $imageSize = $width * $height;

            if ($imageSize > $this->maxImageSize) {
                $error = 'Image is too large for processing. Width x Height should be less than 70 megapixels.';
                throw new ImageTooLargeException($error);
            }
        }
    }

    /**
     * Perform thumbnail image manipulation.
     *
     * @param Image $image The source image.
     * @param string $fit The fit.
     * @param int $width The width.
     * @param int $height The height.
     *
     * @throws \Jcupitt\Vips\Exception
     *
     * @return Image The manipulated image.
     *
     * @SuppressWarnings(PHPMD.NPathComplexity)
     */
    public function doThumbnail(Image $image, string $fit, int $width, int $height): Image
    {
        // Default settings
        $thumbnailOptions = [
            'auto_rotate' => false,
            'linear' => false
        ];

        $isCMYK = $image->interpretation === Interpretation::CMYK;
        $embeddedProfile = Utils::hasProfile($image);

        // Ensure we're using a device-independent color space
        if ($embeddedProfile || $isCMYK) {
            // Embedded profile; fallback in case the profile embedded in the image is broken.
            // No embedded profile; import using default CMYK profile.
            $thumbnailOptions['import_profile'] = $isCMYK ?
                $this->profileMap[Interpretation::CMYK] :
                $this->profileMap[Interpretation::SRGB];

            // Convert to sRGB using embedded or import profile.
            $thumbnailOptions['export_profile'] = $this->profileMap[Interpretation::SRGB];

            // Use "perceptual" intent to better match imagemagick.
            $thumbnailOptions['intent'] = Intent::PERCEPTUAL;
        }

        $inputWidth = $image->width;
        $inputHeight = $image->height;

        $rotation = $this->rotation;
        $swapNeeded = $rotation === 90 || $rotation === 270;

        if ($swapNeeded) {
            // Swap input width and height when rotating by 90 or 270 degrees
            [$inputWidth, $inputHeight] = [$inputHeight, $inputWidth];
        }

        // Scaling calculations
        $targetResizeWidth = $width;
        $targetResizeHeight = $height;

        if ($width > 0 && $height > 0) { // Fixed width and height
            $xFactor = (float)($inputWidth / $width);
            $yFactor = (float)($inputHeight / $height);
            switch ($fit) {
                case 'square':
                case 'squaredown':
                case 'crop':
                    if ($xFactor < $yFactor) {
                        $targetResizeHeight = (int)round($inputHeight / $xFactor);
                    } else {
                        $targetResizeWidth = (int)round($inputWidth / $yFactor);
                    }
                    break;
                case 'letterbox':
                case 'fit':
                case 'fitup':
                    if ($xFactor > $yFactor) {
                        $targetResizeHeight = (int)round($inputHeight / $xFactor);
                    } else {
                        $targetResizeWidth = (int)round($inputWidth / $yFactor);
                    }
                    break;
            }
        } elseif ($width > 0) { // Fixed width
            if ($fit === 'absolute') {
                $targetResizeHeight = $this->h = $inputHeight;
            } else {
                // Auto height
                $yFactor = (float)($inputWidth / $width);
                $this->h = (int)round($inputHeight / $yFactor);

                // Height is missing, replace with a huuuge value to prevent
                // reduction or enlargement in that axis
                $targetResizeHeight = self::VIPS_MAX_COORD;
            }
        } elseif ($height > 0) { // Fixed height
            if ($fit === 'absolute') {
                $targetResizeWidth = $this->w = $inputWidth;
            } else {
                // Auto width
                $xFactor = (float)($inputHeight / $height);
                $this->w = (int)round($inputWidth / $xFactor);

                // Width is missing, replace with a huuuge value to prevent
                // reduction or enlargement in that axis
                $targetResizeWidth = self::VIPS_MAX_COORD;
            }
        } else {
            // Identity transform
            $targetResizeWidth = $this->w = $inputWidth;
            $targetResizeHeight = $this->h = $inputHeight;
        }

        if ($swapNeeded) {
            // Swap target output width and height when rotating by 90 or 270 degrees
            [$targetResizeWidth, $targetResizeHeight] = [$targetResizeHeight, $targetResizeWidth];
        }

        // Assign settings
        $thumbnailOptions['height'] = $targetResizeHeight;

        if ($fit === 'absolute') {
            $thumbnailOptions['size'] = Size::FORCE;
        } else {
            $thumbnailOptions['size'] = $this->withoutEnlargement($fit) ? Size::DOWN : Size::BOTH;
        }

        /**
         * Try to use shrink-on-load for JPEG and WebP, when not
         * applying gamma correction or when trimming isn't required.
         *
         * Note: After this operation the pixel interpretation is sRGB or RGB
         *
         * @see Interpretation::SRGB
         * @see Interpretation::RGB
         */
        return $this->trim !== false || isset($this->gam) ?
            $image->thumbnail_image($targetResizeWidth, $thumbnailOptions) :
            Image::thumbnail($this->tmpFileName, $targetResizeWidth, $thumbnailOptions);
    }
}

File: src/weserv/manipulators/thumbnail.lua
local vips = vips
local utils = require "weserv.helpers.utils"
local error = error
local tonumber = tonumber
local math_floor = math.floor
local ngx_var = ngx.var
local HTTP_NOT_FOUND = ngx.HTTP_NOT_FOUND

-- Profile map to ensure that we use a device-
-- independent color space for the images we process.
local root_dir = ngx_var.weserv_root ~= nil and ngx_var.weserv_root or "."
local profile_map = {
    -- Default sRGB ICC profile from:
    -- https://packages.debian.org/sid/all/icc-profiles-free/filelist
    srgb = root_dir .. "/src/weserv/ICC/sRGB.icm",
    -- Convert to sRGB using default CMYK profile from:
    -- https://www.argyllcms.com/cmyk.icm
    cmyk = root_dir .. "/src/weserv/ICC/cmyk.icm",
}

local VIPS_MAX_COORD = 10000000
local MAX_IMAGE_SIZE = 71000000

--- Thumbnail manipulator
-- @module thumbnail
local manipulator = {}

--- Resolve device pixel ratio.
-- @param dpr The given device pixel ratio.
-- @return The resolved fit.
function manipulator.resolve_dpr(dpr)
    local pixel_ratio = tonumber(dpr)

    -- Pixel ratio may not be nil and needs to be in the range of 0 - 8
    if pixel_ratio ~= nil and pixel_ratio >= 0 and pixel_ratio <= 8 then
        return pixel_ratio
    end

    return 1.0
end

--- Resolve dimension.
-- @param dim The given dimension.
-- @return The resolved dimension.
function manipulator.resolve_dimension(dim)
    local dimension = tonumber(dim)

    -- Dimension may not be nil and not less than 0
    if dimension ~= nil and dimension > 0 then
        return dimension
    end

    return 0
end

--- Resolve fit.
-- @param t The given fit.
-- @return The resolved fit.
function manipulator.resolve_fit(t)
    if t == "fit" or
            t == "fitup" or
            t == "square" or
            t == "squaredown" or
            t == "absolute" or
            t == "letterbox" then
        return t
    end

    if t ~= nil and t:sub(1, 4) == "crop" then
        return "crop"
    end

    return "fit"
end

--- Indicating if we should not enlarge the output image.
-- @param fit The resolved fit.
-- @return bool
function manipulator.without_enlargement(fit)
    return fit == "fit" or fit == "squaredown"
end

--- Should this manipulator process the image?
-- @return Boolean indicating if we should process the image.
function manipulator.should_process(_)
    -- Always process the image via this manipulator.
    return true
end

--- Perform thumbnail image manipulation.
-- @param image The source image.
-- @param args The URL query arguments.
-- @return The manipulated image.
function manipulator.process(image, args)
    if image:width() * image:height() > MAX_IMAGE_SIZE then
        error({
            status = HTTP_NOT_FOUND,
            message = "Image is too large for processing. Width x height should be less than 71 megapixels.",
        })
    end

    -- Resolve target dimensions
    args.w = manipulator.resolve_dimension(args.w)
    args.h = manipulator.resolve_dimension(args.h)

    -- Resolve the device pixel ratio.
    local dpr = manipulator.resolve_dpr(args.dpr)

    -- Apply the device pixel ratio.
    args.w = args.w * dpr
    args.h = args.h * dpr

    local thumbnail_options = {
        auto_rotate = false,
        linear = false,
    }

    local is_cmyk = image:interpretation() == "cmyk"
    local embedded_profile = utils.has_profile(image)

    -- Ensure we're using a device-independent color space
    if embedded_profile or (not embedded_profile and is_cmyk) then
        -- Embedded profile; fallback in case the profile embedded in the image is broken.
        -- No embedded profile; import using default CMYK profile.
        if is_cmyk then
            thumbnail_options.import_profile = profile_map.cmyk
        else
            thumbnail_options.import_profile = profile_map.srgb
        end

        -- Convert to sRGB using embedded or import profile.
        thumbnail_options.export_profile = profile_map.srgb

        -- Use "perceptual" intent to better match imagemagick.
        thumbnail_options.intent = "perceptual"
    end

    local input_width = image:width()
    local input_height = image:height()

    local rotation = args.rotation
    local swap_needed = rotation == 90 or rotation == 270

    if swap_needed then
        input_width, input_height = input_height, input_width
    end

    -- Scaling calculations
    local thumbnail_width = args.w
    local thumbnail_height = args.h

    local fit = manipulator.resolve_fit(args.t)

    local check_max_image_size = true

    if args.w > 0 and args.h > 0 then -- Fixed width and height
        local x_factor = input_width / args.w
        local y_factor = input_height / args.h

        if fit == "square" or fit == "squaredown" or fit == "crop" then
            -- We aim to fill the bounding box, so we must use centre as crop mode
            thumbnail_options.crop = "centre"

            if x_factor < y_factor then
                thumbnail_height = math_floor((input_height / x_factor) + 0.5)
            else
                thumbnail_width = math_floor((input_width / y_factor) + 0.5)
            end
        elseif fit == "letterbox" or fit == "fit" or fit == "fitup" then
            if x_factor > y_factor then
                thumbnail_height = math_floor((input_height / x_factor) + 0.5)
            else
                thumbnail_width = math_floor((input_width / y_factor) + 0.5)
            end
        end
    elseif args.w > 0 then -- Fixed width
        if fit == "absolute" then
            thumbnail_height = input_height
            args.h = input_height
        else
            -- Auto height
            local y_factor = input_width / args.w
            args.h = math_floor((input_height / y_factor) + 0.5)

            -- Height is missing, replace with a huuuge value to prevent
            -- reduction or enlargement in that axis
            thumbnail_height = VIPS_MAX_COORD
        end
    elseif args.h > 0 then -- Fixed height
        if fit == "absolute" then
            thumbnail_width = input_width
            args.w = input_width
        else
            -- Auto width
            local x_factor = input_height / args.h
            args.w = math_floor((input_width / x_factor) + 0.5)

            -- Width is missing, replace with a huuuge value to prevent
            -- reduction or enlargement in that axis
            thumbnail_width = VIPS_MAX_COORD
        end
    else
        -- Identity transform
        thumbnail_width = input_width
        args.w = input_width

        thumbnail_height = input_height
        args.h = input_height

        -- No need to check max image size because
        -- the dimensions are unchanged.
        check_max_image_size = false
    end

    if swap_needed then
        -- Swap target output width and height when rotating by 90 or 270 degrees
        thumbnail_width, thumbnail_height = thumbnail_height, thumbnail_width
    end

    -- Assign settings
    thumbnail_options.height = thumbnail_height

    if fit == "absolute" then
        thumbnail_options.size = "force"
    elseif manipulator.without_enlargement(fit) then
        thumbnail_options.size = "down"

        -- No need to check max image size because
        -- the operation will only downsize.
        check_max_image_size = false
    else
        thumbnail_options.size = "both"
    end

    -- thumbnail_width and thumbnail_height aren't reliable anymore.
    if check_max_image_size and args.w * args.h > MAX_IMAGE_SIZE then
        error({
            status = HTTP_NOT_FOUND,
            message = "Requested image dimensions are too large. Width x height should be less than 71 megapixels.",
        })
    end

    -- Try to use shrink-on-load for JPEG and WebP, when not
    -- applying gamma correction or when trimming isn't required.
    --
    -- Note: After this operation the pixel interpretation is sRGB or RGB
    if args.trim ~= nil or args.gam ~= nil then
        return image:thumbnail_image(thumbnail_width, thumbnail_options)
    else
        return vips.Image.thumbnail(args.tmp_file_name .. args.string_options,
            thumbnail_width, thumbnail_options)
    end
end

return manipulator